addons/website_sale_referral/models/referral_mixin.py

```diff
@@ -0,0 +1,82 @@
+
+
+class MailActivity(models.Model):
+    _inherit = 'mail.activity'
+
+    def _action_done(self, feedback=False, attachment_ids=None):
+        obj = self.env[self.res_model].browse(self.res_id)
+        obj.reward_done = True
+        return super(MailActivity, self)._action_done(feedback=feedback, attachment_ids=attachment_ids)
+
+
+class ReferralMixin(models.AbstractModel):
+    """ Mixin class for objects which can be tracked by referral. """
+    _name = 'referral.mixin'
+    _description = 'Referral Mixin'
+
+    STATES_PRIORITY = {'cancel': 0, 'new': 1, 'in_progress': 2, 'done': 3}
+    REFERRAL_STAGES = {'new': 'New', 'in_progress': 'In Progress', 'done': 'Done'}
+
+    to_reward = fields.Boolean()
+    reward_done = fields.Boolean()
+
+    def get_referral_statuses(self, utm_source_id, referred_email=None):
+        objects = self.find_others(utm_source_id, referred_email)
+
+        result = {}
+        for o in objects:
+            state = o._get_state_for_referral()
+            if(o.referred_email not in result or self.STATES_PRIORITY[state] > self.STATES_PRIORITY[result[o.referred_email]['state']]):
+                result[o.referred_email] = {'state': state, 'name': o.referred_name, 'company': o.referred_company}
+
+        if referred_email:
+            return result.get(referred_email, None)
+        else:
+            return result
+
+    def check_referral_progress(self, old_state, new_state):
+        others_to_reward = self.find_others(self.source_id, referred_email=self.referred_email, extra_criteria=[('to_reward', '=', True)])
+        if new_state != old_state and not len(others_to_reward):
+            self.get_referral_tracking().updates_count += 1
+            if new_state == 'done':
+                self._send_mail('referral_won_email_template', 'Referral won !', {'referred_name': self.referred_name})
+
+                responsible_id = self.env['ir.config_parameter'].sudo().get_param('website_sale_referral.responsible_id') or SUPERUSER_ID
+                if responsible_id:
+                    activity = self.activity_schedule(
+                        act_type_xmlid='website_sale_referral.mail_act_data_referral_reward',
+                        summary='The referrer for this lead deserves a reward',
+                        user_id=responsible_id)
+                    if(callable(getattr(self, 'activity_created', None))):
+                        self.activity_created(activity)
+                    self.to_reward = True
+
+            elif new_state == 'cancel':
+                self._send_mail('referral_cancelled_email_template', 'Referral lost...', {'referred_name': self.referred_name})
+            else:
+                self._send_mail('referral_state_changed_email_template', 'Referral progressed !', {'referred_name': self.referred_name, 'state': _(self.REFERRAL_STAGES[new_state])})
+
+    def find_others(self, utm_source_id, referred_email=None, extra_criteria=[]):
```

Comment:
Thanks for the tip !

The method is now private.

It was renamed into find_others_referrals, but I'm not sure it's a lot more clear since the purpose of this class is to add a layer of abstraction on top of crm_lead and sales_order and the goal of this method is to find other objects with the same utm_source_id.